core/src/avm2/object/xml_object.rs

```diff
@@ -177,6 +177,46 @@ impl<'gc> XmlObject<'gc> {
         Ref::map(self.0.read(), |data| &data.node)
     }
 
+    pub fn contains(&self, value: &Value<'gc>, activation: &mut Activation<'_, 'gc>) -> bool {
+        let node = self.node();
+
+        if let Some(xml) = value.as_object().and_then(|obj| obj.as_xml_object()) {
+            let other = xml.node();
+
+            if node.is_text() || node.is_attribute() {
+                return node.xml_to_string(activation) == other.xml_to_string(activation);
+            }
+
+            if other.is_text() || other.is_attribute() {
+                return false;
+            }
```

Comment:
Any particular reason why we are manually comparing the text nodes and not delegating this to the `equals` method?

Reply:
**Not sure of what I'm saying, I'm not a XML expert**

The default equality check might simply compare the nodes' memory addresses, which would only return true if they are the exact same node. But in the context of XML, we might want to consider two nodes equal if they have the same type (text or attribute) and the same content, even if they are not the same node in memory.
And perhaps this method is faster on large XML documents than using `equal`.

Reply:
The `equals` method you are calling below is defined here: https://github.com/ruffle-rs/ruffle/blob/d1553a203e679d7bb6e8fc37979970087f344f73/core/src/avm2/e4x.rs#L213

This already does the necessary type matching and comparisons.